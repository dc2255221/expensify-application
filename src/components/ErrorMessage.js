import React from 'react';
import { connect } from 'react-redux';

export const ErrorMessage = (props) => (
    <h3> {props.message} </h3>
);

const mapToStateToProps = (state) => (
    {
        message: state.error.message
    }
);

export default connect(mapToStateToProps)(ErrorMessage);
