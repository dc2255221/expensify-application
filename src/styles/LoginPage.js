import styled from 'styled-components';
import { Colors, FontSize, Spacing } from '../constants';

export const BoxLayout = styled.div`
    align-items: center;
    background: url('/images/darkweb.jpg');
    background-size: cover;
    display: flex;
    height: 100vh;
    justify-content: center;
    width: 100vw;
`; 

export const BoxLayoutBox = styled.div`
    background: white;
    border-radius: 3px;
    padding: ${Spacing.l_size} ${Spacing.m_size};
    text-align: center;
    width: 21rem;
    opacity: 0.9;
`;

export const Title = styled.h1`
    margin: 0 0 ${Spacing.m_size} 0;
    line-height: 1;
`; 

export const ButtonContainer = styled.div`
    display: flex;
`;

export const GoogleButton = styled.button`
    background: ${Colors.red};
    border: none;
    color: white;
    font-size: ${FontSize.large};
    font-weight: 300;
    padding: ${Spacing.s_size};
    cursor: pointer;
`;

export const FacebookButton = styled.button`
    background: ${Colors.blue};
    border: none;
    color: white;
    font-size: ${FontSize.large};
    font-weight: 300;
    padding: ${Spacing.s_size};
    cursor: pointer;
`;