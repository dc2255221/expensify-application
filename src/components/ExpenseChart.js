import React from 'react';
import { connect } from 'react-redux';
import { Doughnut } from 'react-chartjs-2';
import selectExpenses from '../selectors/expenses';
import selectTotalByCategory from '../selectors/expenses-category';
import { ChartContainer } from '../styles/ExpenseChart';
import { CatColors } from '../constants';

export const ExpenseChart = ({ expenses }) => {
    const data = {
        datasets: [
            {
                data: [
                    selectTotalByCategory(expenses, 'home-utilities'),
                    selectTotalByCategory(expenses, 'transportation'),
                    selectTotalByCategory(expenses, 'groceries'),
                    selectTotalByCategory(expenses, 'personal-family-care'),
                    selectTotalByCategory(expenses, 'health'),
                    selectTotalByCategory(expenses, 'insurance'),
                    selectTotalByCategory(expenses, 'restaurants-dining'),
                    selectTotalByCategory(expenses, 'shopping-entertainment'),
                    selectTotalByCategory(expenses, 'travel'),
                    selectTotalByCategory(expenses, 'cash-checks-misc'),
                    selectTotalByCategory(expenses, 'giving'),
                    selectTotalByCategory(expenses, 'business-expenses'),
                    selectTotalByCategory(expenses, 'education'),
                    selectTotalByCategory(expenses, 'finance'),
                    selectTotalByCategory(expenses, 'uncategorized')
                ],
                backgroundColor: [
                    CatColors.light_green,
                    CatColors.med_light_green,
                    CatColors.green,
                    CatColors.med_light_green,
                    CatColors.dark_green,
                    CatColors.blue,
                    CatColors.yellow,
                    CatColors.light_orange,
                    CatColors.orange,
                    CatColors.dark_orange,
                    CatColors.pink,
                    CatColors.dark_pink,
                    CatColors.purple,
                    CatColors.dark_purple,
                    CatColors.grey
                ],
                legend: false
            }
        ],
        labels: [
            'Home and Utilities',
            'Transportation',
            'Groceries',
            'Personal and Family Care',
            'Health',
            'Insurance',
            'Restaurants and Dining',
            'Shopping and Entertainment',
            'Travel',
            'Cash, Checks, and Misc',
            'Giving',
            'Business Expenses',
            'Education',
            'Finance',
            'uncategorized'
        ]
    }
    return (
            !!expenses.length &&
            <ChartContainer>
                <Doughnut data={data}/>
            </ChartContainer> 
    )
};

const mapStateToProps = (state) => ({
    expenses: selectExpenses(state.expenses, state.filters),
});

export default connect(mapStateToProps)(ExpenseChart);