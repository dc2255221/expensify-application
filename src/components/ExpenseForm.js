// The dispatching of action onSubmit of form will not happen in this component
// This is because this component needs to be reused in AddExpensePage and EditExpensePage
// Whether we are adding or editing, we will be dispatching different actions
// Hence, we need to pass the data from the submitted form up 
// We do this by calling a prop onSubmit that gets passed in from the parent

import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import { StyledForm, Error, StyledInput, StyledDiv, StyledSelect, StyledTextarea, StyledButton } from '../styles/ExpenseForm';

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            category: props.expense ? props.expense.category : '',
            calendarFocused: false,
            error: ''
        }
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState({ note });
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) { // must be empty or a number with max two decimal spaces
            this.setState({ amount });
        }
    }
    onCategoryChange = (e) => {
        const category = e.target.value;
        this.setState({ category });
    }
    onDateChange = (createdAt) => {
        if (createdAt) { // this will prevent user from being able to clear date value
            this.setState(() => ({ createdAt }));
        }
    }
    onFocusChange = ({ focused }) => {
        this.setState(() => ({calendarFocused: focused}));
    }
    onSubmit = (e) => {
        e.preventDefault(); // prevent full page refresh
        if (!this.state.description || !this.state.amount || !this.state.category) { // validation for description and amount, format check not needed bc we already added regex checking for amount
            this.setState(() => ({ error: 'Please provide description, amount, and category!'}));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                category: this.state.category,
                note: this.state.note   
            });
        }
    }
    render(){
        return (
            <StyledForm onSubmit={this.onSubmit}>
                {this.state.error && <Error>{this.state.error}</Error>}
                <StyledInput
                    type="text"
                    placeholder=" Description"
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <StyledInput
                    type="number"
                    placeholder=" Amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <StyledDiv>
                    <SingleDatePicker
                        id={"date-input"}
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                </StyledDiv>
                <StyledSelect
                    value={this.state.category} 
                    onChange={this.onCategoryChange}
                >
                    <option value="category"> Category </option>
                    <option value="home-and-utilities"> Home and Utilities </option>
                    <option value="personal-and-family-care"> Personal and Family Care </option>
                    <option value="groceries"> Groceries </option>
                    <option value="restaurants-and-dining"> Restaurants and Dining </option>
                    <option value="health"> Health </option>
                    <option value="insurance"> Insurance </option>
                    <option value="transportation"> Transportation </option>
                    <option value="shopping-and-entertainment"> Shopping and Entertainment </option>
                    <option value="travel"> Travel </option>
                    <option value="education"> Education </option>
                    <option value="giving"> Giving </option>
                    <option value="business-expenses"> Business Expenses </option>
                    <option value="finance"> Finance </option>
                    <option value="cash-checks-and-misc"> Cash, Checks, and Misc </option>
                </StyledSelect>
                <StyledTextarea 
                    placeholder="Add a note for your expense (optional)"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                >
                </StyledTextarea>
                <div>
                    <StyledButton> Save Expense </StyledButton>
                </div>
            </StyledForm>
        )
    }
}
