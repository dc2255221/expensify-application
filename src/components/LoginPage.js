import React from 'react';
import { connect } from 'react-redux';
import { startLoginWithGoogle, startLoginWithFacebook, startLoginWithEmailAndPassword } from '../actions/auth';
import { BoxLayout, BoxLayoutBox, Title, StyledForm, StyledInput, ForgotLink, LoginButton, ButtonContainer, GoogleButton, FacebookButton, StyledSpan, Footer, FooterText, SignUpLink, Line } from '../styles/Authentication';

export class LoginPage extends React.Component {
    state = {
        email: '',
        password: '',
        error: ''
    }
    onEmailChange = (e) => {
        const email = e.target.value;
        this.setState({ email });
    }
    onPasswordChange = (e) => {
        const password = e.target.value;
        this.setState({ password });
    }
    onSubmit = (e) => {
        console.log('onSubmit in LoginPage is called');
        e.preventDefault();
        this.props.startLoginWithEmailAndPassword(this.state.email, this.state.password);
    }
    render() {
        return (
            <BoxLayout>
                <BoxLayoutBox>
                    <Title> Log In </Title>
                    <StyledForm onSubmit={this.onSubmit}>
                        {/* {this.state.error && <Error>{this.state.error}</Error>} */}
                        <StyledInput 
                            placeholder="Email" 
                            type="email"
                            autoFocus
                            value={this.state.email}
                            onChange={this.onEmailChange}
                        />
                        <StyledInput 
                            placeholder="Password" 
                            type="text"
                            autoFocus
                            value={this.state.password}
                            onChange={this.onPasswordChange}
                        />
                        <ForgotLink to='/forgot'> Forgot password? </ForgotLink>
                        <LoginButton type="submit" onSubmit={this.onSubmit}> Continue </LoginButton>
                    </StyledForm>
                    <Line/>
                    <ButtonContainer>
                        <FacebookButton onClick={startLoginWithFacebook}> 
                            <img src="/images/facebook.svg" alt="facebook-logo" width="20" height="30"/>
                            <StyledSpan> Continue </StyledSpan>
                        </FacebookButton>
                        <GoogleButton onClick={startLoginWithGoogle}> 
                            <img src="/images/google.svg" alt="google-logo" width="20" height="30"/>
                            <StyledSpan> Continue </StyledSpan> 
                        </GoogleButton>
                    </ButtonContainer>
                    <Line/>
                    <Footer>
                        <FooterText> Don't have an account? </FooterText>
                        <SignUpLink to="/signup"> Sign Up </SignUpLink>
                    </Footer>
                </BoxLayoutBox>
            </BoxLayout>
        )
    }
}

const mapDispatchToProps = (dispatch) => ({
    startLoginWithGoogle: () => dispatch(startLoginWithGoogle()),
    startLoginWithFacebook: () => dispatch(startLoginWithFacebook()),
    startLoginWithEmailAndPassword: (email, password) => dispatch(startLoginWithEmailAndPassword(email, password))
});

export default connect(undefined, mapDispatchToProps)(LoginPage);