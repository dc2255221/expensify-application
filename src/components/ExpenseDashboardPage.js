import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseChart from './ExpenseChart';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpenseSummary';

const ExpenseDashboardPage = () => (
    <div> 
      <ExpenseSummary/>
      <ExpenseChart/>
      <ExpenseListFilters/>
      <ExpenseList/>
    </div>
);

export default ExpenseDashboardPage;