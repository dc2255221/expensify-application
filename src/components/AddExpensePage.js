// The AddExpensePage will dispatch the given action to the redux store
// We need to connect component to store so it can dispatch
// Don't need anything from state so leave first parentheses empty but pass component to second one
// Now, we have access to props.dispatch
// We can call it within our onSubmit eventlistener
// We need addExpense action generator to create the action that we are going to dispatch
// When we did dispatch, we added the data to the redux store
// Now, we need to automatically redirect to Dashboard page onSubmit of form
// Components that we render inside of React.router gets access to many special props
// On history prop, we have a push method which is how we can programatically change pages. The method takes in the string path to Dashboard
// Using it will switch you over as if you clicked Dashboard link through Browser Routing. Hence, no full page refresh

import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
      </div>
    );
  };
};

// way to return dispatcher function, allowing you to abstract them from the components themselves
const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);


