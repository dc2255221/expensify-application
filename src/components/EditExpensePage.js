// We need to connect this component to redux store so we can dispatch action 
// Search the array of expenses for the expense with given id 
// We do need to set up a mapToStateToProps function bc we need to give component access to the expense object
// We will be able to access the state which is great bc that is where expenses array lives
// But, we also need to pass in props as second argument bc we are searching for props.match.params.id
// We will add one new prop expense and will use current props to search expenses array
// Find allows us to search through an array looking for a single item. We determine whether item is found by returning true from the callback
// Now, we will render an instance of ExpenseForm which takes in an onSubmit prop
// We need to make sure data for the given expense is populated within the form
// We do this by passing down the expense as props to ExpenseForm and having it set the values to state 
// If no expense is passed down (i.e. the case of AddExpenseForm), set default values to state 
// Dispatch the action to edit the expense onSubmit of form
// Redirect to the dashboard using history's push method
// Move the removeExpense button from dashboard to EditExpensePage

// React router will render our connected higher order component 
// The higher order component passes the props through and also allows us to add some new ones within the returned object of mapToStateToProps function 

import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/'); 
  }
  onRemove = () => {
    this.props.removeExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  }
  render() {
    return (
        <div>
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.onSubmit}
          />
          <button onClick={this.onRemove}> Remove </button> 
        </div>
    );
  };
};

const mapToStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense( id, expense )),
  removeExpense: (expenseData) => dispatch(removeExpense( expenseData ))
});

export default connect(mapToStateToProps, mapDispatchToProps)(EditExpensePage);