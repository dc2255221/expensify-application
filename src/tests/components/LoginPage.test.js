import React from 'react';
import { LoginPage } from '../../components/LoginPage';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

test('should correctly render Login Page', () => {
    const wrapper = shallow(<LoginPage/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should call startLogin on button click', () => {
    const startLogin = jest.fn();
    const wrapper = shallow(<LoginPage startLogin={startLogin}/>);
    wrapper.find('button').simulate('click');
    expect(startLogin).toHaveBeenCalled();
});