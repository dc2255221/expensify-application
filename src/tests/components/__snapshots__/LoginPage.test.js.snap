// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render Login Page 1`] = `
<styled.div>
  <styled.div>
    <styled.h1>
       Expensify 
    </styled.h1>
    <p>
       It's time to get your expenses under control 
    </p>
    <styled.button
      className="login-button"
    >
       Login with Google 
    </styled.button>
  </styled.div>
</styled.div>
`;
